xchange-coindirect/src/main/java/info/bitrich/xchangestream/coindirect/service/CoindirectStreamingMarketDataService.java

```diff
@@ -0,0 +1,144 @@
+import info.bitrich.xchangestream.coindirect.CoindirectStreamingService;
+import info.bitrich.xchangestream.coindirect.dto.CoindirectAccountOrderEvent;
+import info.bitrich.xchangestream.coindirect.dto.CoindirectEvent;
+import info.bitrich.xchangestream.coindirect.dto.CoindirectOrderBookEvent;
+import info.bitrich.xchangestream.coindirect.dto.CoindirectTradeEvent;
+import info.bitrich.xchangestream.core.StreamingMarketDataService;
+import io.reactivex.Observable;
+import io.reactivex.functions.Consumer;
+import org.knowm.xchange.Exchange;
+import org.knowm.xchange.coindirect.CoindirectAdapters;
+import org.knowm.xchange.coindirect.dto.account.CoindirectAccountChannel;
+import org.knowm.xchange.coindirect.dto.marketdata.CoindirectOrderbook;
+import org.knowm.xchange.coindirect.service.CoindirectAccountServiceRaw;
+import org.knowm.xchange.coindirect.service.CoindirectMarketDataService;
+import org.knowm.xchange.coindirect.service.CoindirectMarketDataServiceRaw;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.Order;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.knowm.xchange.dto.marketdata.OrderBookUpdate;
+import org.knowm.xchange.dto.marketdata.Ticker;
+import org.knowm.xchange.dto.marketdata.Trade;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CoindirectStreamingMarketDataService implements StreamingMarketDataService {
+    private static final Logger LOG = LoggerFactory.getLogger(CoindirectStreamingMarketDataService.class);
+
+    private final CoindirectStreamingService service;
+    private final Exchange exchange;
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private Map<CurrencyPair, CoindirectOrderbook> coindirectOrderbookMap = new HashMap<>();
+    private Map<CurrencyPair, OrderBook> orderBookMap = new HashMap<>();
+    private CoindirectMarketDataServiceRaw coindirectMarketDataServiceRaw;
+    private CoindirectAccountServiceRaw coindirectAccountServiceRaw;
+
+    public CoindirectStreamingMarketDataService(CoindirectStreamingService service, Exchange exchange) {
+        this.service = service;
+        this.exchange = exchange;
+    }
+
+    private CoindirectMarketDataServiceRaw getCoindirectMarketDataServiceRaw() {
+        if(coindirectMarketDataServiceRaw == null) {
+            coindirectMarketDataServiceRaw = new CoindirectMarketDataServiceRaw(exchange);
+        }
+        return coindirectMarketDataServiceRaw;
+    }
+
+    private CoindirectAccountServiceRaw getCoindirectAccountServiceRaw() {
+        if(coindirectAccountServiceRaw == null) {
+            coindirectAccountServiceRaw = new CoindirectAccountServiceRaw(exchange);
+        }
+        return coindirectAccountServiceRaw;
+    }
+
+    @Override
+    public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
```

Comment:
This code permits multiple parallel subscriptions to the same channel but is not thread safe as a result of shared state